include/gridtools/sid/loop.hpp

```diff
@@ -637,6 +638,30 @@ namespace gridtools {
             return {};
         }
 
+        template <class Key, int UnrollFactor, class NumSteps, class Step = integral_constant<int, 1>>
+        constexpr GT_FUNCTION auto make_unrolled_loop(NumSteps num_steps, Step step = {}) {
+            using u = integral_constant<int, UnrollFactor>;
+            return [step,
+                       unrolled = make_loop<Key>(num_steps / u(), step * u()),
+                       epilogue = make_loop<Key>(num_steps % u(), step),
+                       epilogue_start = step * ((num_steps / u()) * u())](auto &&fun) {
+                return [unrolled = unrolled([step, fun=std::forward<decltype(fun)>(fun)](auto &&ptr, auto const strides) {
+                    ::gridtools::host_device::for_each<meta::make_indices_c<UnrollFactor>>([&](auto) {
+                        fun(std::forward<decltype(ptr)>(ptr), strides);
+                        shift(std::forward<decltype(ptr)>(ptr), get_stride<Key>(strides), step);
+                    });
+                    shift(std::forward<decltype(ptr)>(ptr), get_stride<Key>(strides), -step * u());
```

Comment:
Do you remember why the loop takes ptr possibly by reference and therefore has to do this back shift?

Reply:
The pointer is taken by reference and back shifts are implemented in all SID loop variants. Not sure what the reason for the design decision was, maybe to save memory/registers on large SID composites as otherwise we would always copy them?